webapp/pages/admin/users.vue

```diff
@@ -1,15 +1,173 @@
+        <ds-flex-item width="50px">
+          <ds-button @click="next" :disabled="!hasNext" icon="arrow-right" primary />
+        </ds-flex-item>
+        <ds-flex-item width="50px">
+          <ds-button @click="back" :disabled="!hasPrevious" icon="arrow-left" primary />
+        </ds-flex-item>
+      </ds-flex>
+    </ds-card>
+    <ds-card v-else>
+      <ds-placeholder>
+        {{ $t('admin.users.empty') }}
+      </ds-placeholder>
+    </ds-card>
+  </div>
 </template>
 
 <script>
-import HcEmpty from '~/components/Empty.vue'
+import gql from 'graphql-tag'
+import isemail from 'isemail'
 
 export default {
-  components: {
-    HcEmpty,
+  data() {
+    const pageSize = 15
+    return {
+      offset: 0,
+      pageSize,
+      first: pageSize,
+      User: [],
+      hasNext: false,
+      email: null,
+      filter: null,
+      form: {
+        formData: {
+          query: '',
+        },
+      },
+    }
+  },
+  computed: {
+    hasPrevious() {
+      return this.offset > 0
+    },
+    fields() {
+      return {
+        index: '#',
+        name: this.$t('admin.users.table.columns.name'),
+        slug: this.$t('admin.users.table.columns.slug'),
+        createdAt: this.$t('admin.users.table.columns.createdAt'),
+        contributionsCount: {
+          label: '🖉',
+          align: 'right',
+        },
+        commentedCount: {
+          label: '🗨',
+          align: 'right',
+        },
+        shoutedCount: {
+          label: '❤',
+          align: 'right',
+        },
+        role: {
+          label: this.$t('admin.users.table.columns.role'),
+          align: 'right',
+        },
+      }
+    },
+  },
+  apollo: {
+    User: {
+      query() {
+        return gql(`
+        query($filter: _UserFilter, $first: Int, $offset: Int, $email: String) {
+          User(email: $email, filter: $filter, first: $first, offset: $offset, orderBy: createdAt_desc) {
+            id
+            name
+            slug
+            role
+            createdAt
+            contributionsCount
+            commentedCount
+            shoutedCount
+          }
+        }
+      `)
+      },
+      variables() {
+        const { offset, first, email, filter } = this
+        const variables = { first, offset }
+        if (email) variables.email = email
+        if (filter) variables.filter = filter
+        return variables
+      },
+      update({ User }) {
+        if (!User) return []
+        this.hasNext = User.length >= this.pageSize
+        if (User.length <= 0 && this.offset > 0) return this.User // edge case, avoid a blank page
+        return User.map((u, i) => Object.assign({}, u, { index: this.offset + i }))
+      },
+    },
+  },
+  methods: {
+    back() {
+      this.offset = Math.max(this.offset - this.pageSize, 0)
+    },
+    next() {
+      this.offset += this.pageSize
+    },
+    submit(formData) {
+      this.offset = 0
+      const { query } = formData
+      if (isemail.validate(query)) {
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-15T19:25:02Z" title="Monday, July 15th 2019, 9:25:02 pm +02:00">Jul 15, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,15 +1,173 @@
 <template>
-  <ds-card :header="$t('admin.users.name')">
-    <hc-empty icon="tasks" message="Coming Soon…" />
-  </ds-card>
+  <div>
+    <ds-space>
+      <ds-card :header="$t('admin.users.name')">
+        <ds-form v-model="form" @submit="submit">
+          <ds-flex gutter="small">
+            <ds-flex-item width="90%">
+              <ds-input
+                model="query"
+                :placeholder="$t('admin.users.form.placeholder')"
+                icon="search"
+              />
+            </ds-flex-item>
+            <ds-flex-item width="30px">
+              <ds-button primary type="submit" icon="search" :loading="$apollo.loading" />
+            </ds-flex-item>
+          </ds-flex>
+        </ds-form>
+      </ds-card>
+    </ds-space>
+    <ds-card v-if="User && User.length">
+      <ds-table :data="User" :fields="fields" condensed>
+        <template slot="index" slot-scope="scope">
+          {{ scope.row.index }}.
+        </template>
+        <template slot="name" slot-scope="scope">
+          <nuxt-link
+            :to="{
+              name: 'profile-id-slug',
+              params: { id: scope.row.id, slug: scope.row.slug },
+            }"
+          >
+            <b>{{ scope.row.name | truncate(20) }}</b>
+          </nuxt-link>
+        </template>
+        <template slot="slug" slot-scope="scope">
+          <nuxt-link
+            :to="{
+              name: 'profile-id-slug',
+              params: { id: scope.row.id, slug: scope.row.slug },
+            }"
+          >
+            <b>{{ scope.row.slug | truncate(20) }}</b>
+          </nuxt-link>
+        </template>
+        <template slot="createdAt" slot-scope="scope">
+          {{ scope.row.createdAt | dateTime }}
+        </template>
+      </ds-table>
+      <ds-flex direction="row-reverse">
+        <ds-flex-item width="50px">
+          <ds-button @click="next" :disabled="!hasNext" icon="arrow-right" primary />
+        </ds-flex-item>
+        <ds-flex-item width="50px">
+          <ds-button @click="back" :disabled="!hasPrevious" icon="arrow-left" primary />
+        </ds-flex-item>
+      </ds-flex>
+    </ds-card>
+    <ds-card v-else>
+      <ds-placeholder>
+        {{ $t('admin.users.empty') }}
+      </ds-placeholder>
+    </ds-card>
+  </div>
 </template>
 
 <script>
-import HcEmpty from '~/components/Empty.vue'
+import gql from 'graphql-tag'
+import isemail from 'isemail'
 
 export default {
-  components: {
-    HcEmpty,
+  data() {
+    const pageSize = 15
+    return {
+      offset: 0,
+      pageSize,
+      first: pageSize,
+      User: [],
+      hasNext: false,
+      email: null,
+      filter: null,
+      form: {
+        formData: {
+          query: '',
+        },
+      },
+    }
+  },
+  computed: {
+    hasPrevious() {
+      return this.offset > 0
+    },
+    fields() {
+      return {
+        index: '#',
+        name: this.$t('admin.users.table.columns.name'),
+        slug: this.$t('admin.users.table.columns.slug'),
+        createdAt: this.$t('admin.users.table.columns.createdAt'),
+        contributionsCount: {
+          label: '🖉',
+          align: 'right',
+        },
+        commentedCount: {
+          label: '🗨',
+          align: 'right',
+        },
+        shoutedCount: {
+          label: '❤',
+          align: 'right',
+        },
+        role: {
+          label: this.$t('admin.users.table.columns.role'),
+          align: 'right',
+        },
+      }
+    },
+  },
+  apollo: {
+    User: {
+      query() {
+        return gql(`
+        query($filter: _UserFilter, $first: Int, $offset: Int, $email: String) {
+          User(email: $email, filter: $filter, first: $first, offset: $offset, orderBy: createdAt_desc) {
+            id
+            name
+            slug
+            role
+            createdAt
+            contributionsCount
+            commentedCount
+            shoutedCount
+          }
+        }
+      `)
+      },
+      variables() {
+        const { offset, first, email, filter } = this
+        const variables = { first, offset }
+        if (email) variables.email = email
+        if (filter) variables.filter = filter
+        return variables
+      },
+      update({ User }) {
+        if (!User) return []
+        this.hasNext = User.length >= this.pageSize
+        if (User.length <= 0 && this.offset > 0) return this.User // edge case, avoid a blank page
+        return User.map((u, i) => Object.assign({}, u, { index: this.offset + i }))
+      },
+    },
+  },
+  methods: {
+    back() {
+      this.offset = Math.max(this.offset - this.pageSize, 0)
+    },
+    next() {
+      this.offset += this.pageSize
+    },
+    submit(formData) {
+      this.offset = 0
+      const { query } = formData
+      if (isemail.validate(query)) {
```


I guess you explored the option of using our styleguides' async validators? https://github.com/yiminghe/async-validator
and maybe couldn't because it doesn't have to be an email, it can be other input as well?

`isemail` is an interesting package, it's used by more than 450k people, but only 189 stars :woman_shrugging:

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-07-15T21:00:06Z" title="Monday, July 15th 2019, 11:00:06 pm +02:00">Jul 15, 2019</time>_

---


Yeah, I found `isemail` in our `yarn.lock` when I  `git grep 'email'`. You're right: I want to check if it's an email and I don't want to throw an error or sth. in case it's not an email.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-15T21:28:31Z" title="Monday, July 15th 2019, 11:28:31 pm +02:00">Jul 15, 2019</time>_

---


great! thanks for walking us through the process! That is a great way to decide when to add a package or not :smile_cat: